sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/NumberUtilsSuite.scala

```diff
@@ -0,0 +1,273 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.util
+
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.util.NumberUtils.{format, parse}
+import org.apache.spark.sql.types.Decimal
+import org.apache.spark.unsafe.types.UTF8String
+
+class NumberUtilsSuite extends SparkFunSuite {
+
+  private def failParseWithInvalidInput(
+      input: UTF8String, numberFormat: String, errorMsg: String): Unit = {
+    val e = intercept[IllegalArgumentException](parse(input, numberFormat))
+    assert(e.getMessage.contains(errorMsg))
+  }
+
+  private def failParseWithAnalysisException(
+      input: UTF8String, numberFormat: String, errorMsg: String): Unit = {
+    val e = intercept[AnalysisException](parse(input, numberFormat))
+    assert(e.getMessage.contains(errorMsg))
+  }
+
+  private def failFormatWithAnalysisException(
+      input: Decimal, numberFormat: String, errorMsg: String): Unit = {
+    val e = intercept[AnalysisException](format(input, numberFormat))
+    assert(e.getMessage.contains(errorMsg))
+  }
+
+  test("parse") {
+    failParseWithInvalidInput(UTF8String.fromString("454"), "",
+      "Format '' used for parsing string to number or formatting number to string is invalid")
+
+    // Test '9' and '0'
+    failParseWithInvalidInput(UTF8String.fromString("454"), "9",
+      "Format '9' used for parsing string to number or formatting number to string is invalid")
+    failParseWithInvalidInput(UTF8String.fromString("454"), "99",
+      "Format '99' used for parsing string to number or formatting number to string is invalid")
+
+    assert(parse(UTF8String.fromString("454"), "999") === Decimal(454))
+    assert(parse(UTF8String.fromString("054"), "999") === Decimal(54))
+    assert(parse(UTF8String.fromString("404"), "999") === Decimal(404))
+    assert(parse(UTF8String.fromString("450"), "999") === Decimal(450))
+    assert(parse(UTF8String.fromString("454"), "9999") === Decimal(454))
+    assert(parse(UTF8String.fromString("054"), "9999") === Decimal(54))
+    assert(parse(UTF8String.fromString("404"), "9999") === Decimal(404))
+    assert(parse(UTF8String.fromString("450"), "9999") === Decimal(450))
```

Comment:
Could write this and below  in more general way like:
```scala
    Seq(
      ("454", "999") -> Decimal(454)
    ).foreach { case ((str, format), expected) =>
      assert(parse(UTF8String.fromString(str), format) === expected)
    }
```

Reply:
OK